plugins/net_plugin/include/eos/net_plugin/message_buffer.hpp

```diff
@@ -0,0 +1,278 @@
+    }
+
+    /*
+     *  Adds an additional buffer of buffer_len to the chain of buffers.
+     *  Does not affect the read or write pointer.
+     */
+    void add_buffer_to_chain() {
+      buffers.push_back(pool().malloc());
+    }
+
+    /*
+     *  Adds additional buffers of length buffer_len to the chain of buffers
+     *  in order to add at least bytes to the space available.
+     *  Does not affect the read or write pointer.
+     */
+    void add_space(uint32_t bytes) {
+      int buffers_to_add = bytes / buffer_len + 1;
+      for (int i = 0; i < buffers_to_add; i++) {
+        buffers.push_back(pool().malloc());
+      }
+    }
+
+    /*
+     *  Returns the current number of bytes remaining to be read.
+     *  Logically, this is the different between where the read and write pointers are.
+     */
+    uint32_t bytes_to_read() const {
+      return (write_ind.first - read_ind.first) * buffer_len + write_ind.second - read_ind.second;
+    }
+
+    /*
+     *  Returns the current number of bytes available to be written.
+     *  Logically, this is the different between the write pointer and the
+     *  end of the buffer.  If this is not enough room, call either
+     *  add_buffer_to_chain() or add_room() to create more space.
+     */
+    uint32_t bytes_to_write() const {
+      return total_bytes() - write_ind.first * buffer_len - write_ind.second;
+    }
+
+    /*
+     *  Returns the total number of bytes in the buffer chain.
+     */
+    uint32_t total_bytes() const {
+      return buffer_len * buffers.size();
+    }
+
+    /*
+     *  Advances the read pointer/index the supplied number of bytes along
+     *  the buffer chain.  Any buffers that the read pointer moves beyond
+     *  will be removed from the buffer chain.  If the read pointer becomes
+     *  equal to the write pointer, the message buffer will be reset to
+     *  its initial state (one buffer with read and write pointers at the
+     *  start).
+     */
+    void advance_read_ptr(uint32_t bytes) {
+      advance_index(read_ind, bytes);
+      if (read_ind == write_ind) {
+        read_ind = { 0, 0 };
+        write_ind = { 0, 0 };
+        while (buffers.size() > 1) {
+          pool().destroy(buffers.back());
+          buffers.pop_back();
+        }
+      } else if (read_ind.first > 0) {
+        while (read_ind.first > 0) {
+          pool().destroy(buffers.front());
+          buffers.pop_front();
+          read_ind.first--;
+          write_ind.first--;
+        }
+      }
+    }
+
+    /*
+     *  Advances the write pointer/index the supplied number of bytes along
+     *  the buffer chain.
+     */
+    void advance_write_ptr(uint32_t bytes) {
+      advance_index(write_ind, bytes);
+    }
+
+    /*
+     *  Debug function that will print the state of the message buffer
+     *  to standard out.
+     */
+    void show_state() const {
+      std::cout << "Number of buffers = " << buffers.size() << "\n"
```

Comment:
Should use the logger like in net_plugin.

Reply:
I was using this member function to debug some of the unit tests, I couldn't decide whether to leave it in or not. I might just take it out.

Reply:
I agree with Kevin on this. We need to be consistent with our logging strategy. No other plugin or library code uses std::cout or std::cerr.

Reply:
I agree with Kevin. We need to be consistent with our logging.